Canary test workflow, checks that key downstream packages still build

ChristianMurphy opened the pull request:
<!--
  Please check the needed checkboxes ([ ] -> [x]). Leave the
  comments as they are, they won’t show on GitHub.
  We are excited about pull requests, but please try to limit the scope, provide
  a general description of the changes, and remember, it’s up to you to convince
  us to land it.
-->

### Initial checklist

*   [x] I read the support docs <!-- https://github.com/micromark/.github/blob/main/support.md -->
*   [x] I read the contributing guide <!-- https://github.com/micromark/.github/blob/main/contributing.md -->
*   [x] I agree to follow the code of conduct <!-- https://github.com/micromark/.github/blob/main/code-of-conduct.md -->
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) <!-- https://github.com/search?q=user%3Amicromark&type=Issues -->
*   [x] If applicable, I’ve added docs and tests

### Description of changes

Follow up to: https://github.com/unifiedjs/ideas/issues/7

> Unified is a large and highly connected community of packages.
> Changes to any package could have impact downstream packages.
> Currently anticipating impact is a purely manual process.

> It could be helpful to use downstream packages, which themselves have comprehensive test suites, to help automatically detect what changes may be breaking or have unintended impacts.

This aims to help alleviate some of that risk by choosing a few key packages.
* mdast-util-from-markdown
* remark
* react-markdown

which build on this, and have heavy usage, and comprehensive test suites, to anticipate potential downstream issues faster.

<!--do not edit: pr-->

ChristianMurphy commented on .github/workflows/canary.yml:
```diff
@@ -0,0 +1,63 @@
+      - name: test mdast-util-from-markdown
+        run: |
+          npm install
+          npx rimraf "**/micromark"
+          npx rimraf "**/micromark-util-*"
```
this removes the pinned versions of these libraries that npm installed, meaning node will continue searching upward, until it finds the parent `micromark/node_modules` folder with the just-built version of micromark.

ChristianMurphy commented:
Good news: this works!
Bad news: there appear to be failing tests upstream

wooorm commented on .github/workflows/canary.yml:
```diff
@@ -0,0 +1,49 @@
+      - name: test ${{matrix.package}}
+        run: |
+          npm install
+          npx rimraf "**/micromark"
```
Wouldn’t `npx` significantly slow down this process compared to having it (guessing we already have it?)

ChristianMurphy replied:
We already have it, `npx` is allowing us to call it without needing to specify which local `node_modules` folder it is in.

wooorm reviewed (commented):
Fixing the CI failure is most important, otherwise, good idea!

wooorm commented:
Perhaps something to not run this on every PR? Similar to that “T-Shaped” model from MDX?

ChristianMurphy commented:
> Perhaps something to not run this on every PR?

I think there would be value in running this on every PR, seeing if code changes impact upstream.
And even if changes aren't code related, it can be an early warning for unrelated upstream issues.
This set of checks itself probably wouldn't be blocking for PRs.

> Similar to that “T-Shaped” model from MDX?

There the T-Shaped model has to do with the matrix of jobs run, not when it is run, right?
I did drop windows from the matrix for now, because it appears most upstream projects do not test on Windows.
Which (may) eliminate the need to a T-Shaped model.

wooorm commented on .github/workflows/canary.yml:
```diff
@@ -0,0 +1,48 @@
+      matrix:
+        os:
+          - ubuntu-latest
+        node:
+          - lts/erbium
+          - node
```
I see no strong benefit in trying to test Erbium, assuming that in almost all cases the result would be the same as on `node`?

wooorm commented on .github/workflows/canary.yml:
```diff
@@ -0,0 +1,48 @@
+        uses: dcodeIO/setup-node-nvm@master
+        with:
+          node-version: ${{matrix.node}}
+      - name: update npm
+        run: npm install -g npm
+      - name: install micromark
+        run: npm install
+      - name: build micromark
+        run: npm run build
```
```suggestion
      - name: micromark
        run: npm install -g npm && npm install && npm run build
```

This should likely work and be uninteresting. And if it doesn’t work, the other CIs should fail

wooorm replied:
Can also use `|`s for multilines, like the next item

wooorm commented on .github/workflows/canary.yml:
```diff
@@ -0,0 +1,48 @@
+      - name: test ${{matrix.package}}
+        run: |
+          npm install
+          npx rimraf "**/micromark"
+          npx rimraf "**/micromark-util-*"
```
What is being removed here? Are there `micromark` files or folders in `micromark/canary/remarkjs/remark/`?
Why just `micromark` itself and `micromark-util-*`, but not other `micromark-*` files/folders?

ChristianMurphy replied:
copying from earlier in the discussion https://github.com/micromark/micromark/pull/106#discussion_r767339359

> this removes the pinned versions of these libraries that npm installed, meaning node will continue searching upward, until it finds the parent `micromark/node_modules` folder with the just-built version of micromark.

---

> Are there micromark files or folders in micromark/canary/remarkjs/remark/?

There are, this runs right after `npm install` in the _micromark/canary/remarkjs/remark_ folder, so it has a frozen/published copy of micromark installed, we want it to find the latest, just built, copy instead.

> Why just micromark itself and micromark-util-*, but not other micromark-* files/folders?

It seemed to work, open to suggestions for a better remove pattern to get all the micromark packages in this repo cleared at once.

wooorm replied:
ahhh, forgot that I already asked it.
Wouldn’t `npx rimraf "node_modules/micromark*"` do the same but catch everything in [`packages/`](https://github.com/micromark/micromark/tree/main/packages)? (or `npx rimraf "node_modules/{micromark,micromark*}"` if the asterisk matches one or more, which I don’t think it does?)

ChristianMurphy replied:
One possible alternative, something like

```
rm -rf `npm ls --workspaces --json | jq "[.dependencies | keys[]] | \"**/\" + .[]"`
```

which tries to get at the parent workspace, list the packages installed there, create a `**/{package name}` pattern for each, then passing all that back to `rm -rf`

wooorm replied:
Hmm, but what about packages needed in the canary packages though? `micromark-extension-mdx`?
Problem is that this repo houses `micromark-factory-*` too, but those extensions can also use `micromark-*` packages? 🤔

ChristianMurphy replied:
> Wouldn’t `npx rimraf "node_modules/micromark*"` do the same but catch everything in packages/?

it would! but, it may be too aggressive, also removing things like `micromark-extension-gfm` which may be needed by unit tests and not part of the parent micromark repo.

wooorm replied:
Would `rm -rf "canary/.../node_modules/${ls packages/}"` do the same, but cleaner?

ChristianMurphy replied:
which admittedly can even happen with my current pattern.
For example

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'micromark-util-events-to-acorn' imported from /home/runner/work/micromark/micromark/canary/micromark/micromark-extension-mdx/node_modules/micromark-factory-mdx-expression/dev/index.js
    at new NodeError (node:internal/errors:371:5)
    at packageResolve (node:internal/modules/esm/resolve:864:9)
    at moduleResolve (node:internal/modules/esm/resolve:910:18)
    at defaultResolve (node:internal/modules/esm/resolve:1005:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

https://github.com/micromark/micromark/runs/4627735817?check_suite_focus=true#step:6:36

ChristianMurphy replied:
> Would `rm -rf "canary/.../node_modules/${ls packages/}"` do the same, but cleaner?

potentially, I think we want the reverse of this though. :thinking: 
inside the canary folder, we want to list the packages in the parent micromark folder, and use that list to remove frozen/published copies in the canary folder.

wooorm replied:
Huh, that’s what my suggestion does, right?
- List all packages maintained in the micromark monorepo (which are hoisted to the monorepo `node_modules/`?)
- Install an example repo normally, separately from the monorepo
- Remove what monorepo packages from the example repo `node_modules`

wooorm replied:
Though my example was assuming a CWD of the monorepo, from an example repo my example would be, say, `rm -rf "./node_modules/${ls ../../../packages/}"`

ChristianMurphy replied:
testing locally I get `bad substitution`.
But parsing out your intent.

This script is running inside the canary/{package-name} folder.
so `ls packages` would list package inside the `canary` folder right?
`canary` would then try to go into a folder that doesn't exist.
`...` would take it up two levels to the parent micromark folder
`node_modules` would go into the parent's `node_modules` directory

ChristianMurphy replied:
resolved in  42e5976

wooorm commented on .github/workflows/canary.yml:
```diff
@@ -0,0 +1,48 @@
+  - push
+jobs:
+  canary:
+    name: testing ${{matrix.package}} with ${{matrix.node}} on ${{matrix.os}}
```
(see later for more info, but: I think testing only on `ubuntu-latest` and `node` is already useful, and I don’t see much added benefits by checking older node too).

```suggestion
    name: ${{matrix.package}}
```

Name is quite long and can’t be displayed properly (sidebar here: https://github.com/micromark/micromark/runs/4620836214?check_suite_focus=true)

wooorm commented:
Looking at the, how do you call this, build preview, it looks chaotic imo.
- Could the canary/* stuff be ordered after main/*?
- `canary / testing / $user/$project` -> `canary-in-the-gold-mine / $package` or `citgm / $package` maybe?

<img width="924" alt="Screen Shot 2021-12-24 at 16 46 09" src="https://user-images.githubusercontent.com/944406/147362980-18194411-ff49-43e6-b713-6ca058be3841.png">

ChristianMurphy commented:
> Could the canary/* stuff be ordered after main/*?

I'm not sure if it can.
The order appears to be non-alphabetical.

> `canary / testing / $user/$project` -> `canary-in-the-gold-mine / $package or citgm` / `$package` maybe?

Happy to rename things if wanted, I'm not sure it will fix the order issue though.

wooorm commented:
> I'm not sure if it can. The order appears to be non-alphabetical.

Yeah, though, taking MDX as an example: https://github.com/mdx-js/mdx/pull/1854, it seems to indicate that different YAML files result in unordered, but when inside the same file, it does work?

> Happy to rename things if wanted, I'm not sure it will fix the order issue though.

I meant the points separately. I guess `$packageName` doesn’t make sense though, but `$user/$project` does, as this seems to be more about the repo rather the npm package?

ChristianMurphy commented:
> but when inside the same file, it does work?

Moved canary job into main.yml

> as this seems to be more about the repo rather the npm package?

it is more about the repositories, updated naming to better reflect this
